### ISSUE TITLE
Summarise draws is probibitively slow for large number of parameters

### ISSUE THREAD

**Issue by spinkney** [stan-dev/posterior#98]
I updated `cmdstanr` to use `data.table` and then noticed that the second bottleneck is the `posterior` package. For small number of parameters `data.table` isn't much different but once the number of parameters grows to 10s or 100s of thousands the `summarise_draws()` function becomes much, much slower.

For example, here is a benchmark of 40901 parameter summary using mean, median, and mad of the posterior `summarise_draws` vs a custom function using `data.table`. The difference is 5x slower for posterior version.
```r
# 40901 parameters
> mbm <- microbenchmark(
+   ex_posterior_summary = summarise_draws(ex_posterior, mean, median, mad),
+   ex_dt_summary = summarise_draws_dt(ex_posterior),
+   times = 10L
+ )
> mbm
Unit: seconds
                 expr       min        lq      mean    median        uq       max neval
 ex_posterior_summary 174.41825 175.31964 178.12328 177.51151 180.00368 185.84406    10
        ex_dt_summary  33.50971  33.85208  34.53455  34.47042  34.86485  35.95574    10
> 178/34
[1] 5.235294
```
When I increase the parameters to 79436. The difference is now 6.67x slower. 
```r
dim(ex_posterior)
[1]  1000 79436
> mbm
Unit: seconds
                 expr       min        lq      mean    median        uq       max neval
 ex_posterior_summary 541.82929 541.82929 541.82929 541.82929 541.82929 541.82929     1
        ex_dt_summary  81.64224  81.64224  81.64224  81.64224  81.64224  81.64224     1
> 541/81
[1] 6.679012
```
The function `summarise_draws_dt` I created using `data.table` is
```r
summarise_draws_dt <-function(d) {
    ex_dt <- as.data.table(d)
    dt_sum <- t(ex_dt[, as.list(unlist(lapply(.SD, function(x) list(mean = mean(x), 
                                                                    median = median(x),
                                                                    # sd = sd(x),
                                                                    mad = mad(x)
                                                                ))))])
    dt_sum <- as.data.table(dt_sum, keep.rownames = T)
    dt_sum[, c("variable", "cast") := tstrsplit(rn, ".", fixed=TRUE)]
    return(dcast(dt_sum, variable ~ cast, value.var = "V1", fun.aggregate = mean))
}
````

Equivalency of output
```r
# output is equivalent
# > summarise_draws
# A tibble: 14 x 4
#   variable           mean  median     mad
#   <chr>             <dbl>   <dbl>   <dbl>
# 1 X_beta_raw[1,1]  2.26    2.26   0.00497

# > summarise_draws_dt
#           variable         mad        mean      median
#   1: X_beta_raw[1,1] 0.004974123  2.26025694  2.26014500
```

**Comment by rok-cesnovar** [comment]
I am wondering if this is also related to why `as_draws_array()` seems to be slow. For example, reading 2000 samples for 50k parameters from CSV into R takes 4.5s on my system. But then doing as_draws_array on that takes ~3s. That seems a lot, especially comparing that to actually reading in a file. I have not yet looked into it more deeply though.

**Comment by jsocolar** [comment]
If nobody's already on this, I think I'll soon have some time to give this one a crack.

Four questions though:
1. Would a solution that just deals with `summarise_draws()` be welcome, or has this issue exposed other concerns (e.g. in `as_draws_array()` that you'd prefer to deal with all at once in a more fundamental way?
2. I assume that's fine to introduce a dependency on data.table since cmdstanr shares this dependency? If not there's a base R workaround (albeit likely an inferior one); see issue #104.
3. Is there interest in also supporting the ability to optionally compute the summaries in parallel, and if so if there are any strong preferences about what package to use for that?
4. What kind of performance testing and/or edge-case testing would you like to see for something like this prior to merging?

**Comment by tjmahr** [comment]
future seems like the best way to do parallelism because it abstracts over many different backends.

**Comment by spinkney** [comment]
@jsocolar , what is the time difference using `summarise_draws_dt` referenced above and your parallelized solution in issue #104? 

One nice thing about data.table is that they continue to parallelize internally. See https://www.rdocumentation.org/packages/data.table/versions/1.13.0/topics/setDTthreads.

**Comment by paul-buerkner** [comment]
Thank you all for the interesting suggestions and discussion. I think we should be quite careful about adding dependencies but I agree it could be worthwhile here. Perhaps we could use a 'suggest' option, which could be activated by the user in which case any speedup package would actually be required.

**Comment by jsocolar** [comment]
@spinkney I'll try to benchmark sometime in the next couple of days.  

Do you have a sense of which (if any) parts of your `summarise_draws_dt()` function are amenable to the parallelization available within data.table? I assume that there's no way that data.table is able to voodoo its way into parallelizing the computation that occurs in your `lapply()` statement, which I imagine is the main rate-limiting step in your code, and could become nontrivial for really big models (I'm wrangling a 213K parameter model at the moment).

**Comment by avehtari** [comment]
I'll add and example where draws object has 29660 variables, and variable f is a vector of length 7295.
```
fit$summary("f","median")
```
takes 11s, while
```
apply(as_draws_matrix(fit$draws("f")), 2, median)
```
takes only 1.1s.

**Comment by paul-buerkner** [comment]
Just a quick note that some speedup was achieved as part of fixing #119. Still not satisfactory though.

**Comment by jsocolar** [comment]
Ok, so better late than never I guess.
I ended up not deeply pursuing @spinkney's data.table solution (which I still suspect is fastest) because I didn't want to deal with rewriting `rhat()` and other functions that require accounting for the chain ids to deal with data.table objects.  Perhaps this would still be the way to go... I dunno.

Anyhow, it's still the case that
```
parsummarise_draws <- function(draws, n_cores=1, n_chunks=NULL){
  n_pars <- dim(draws)[3]
  if(is.null(n_chunks)){n_chunks <- ceiling(n_pars/500)}
  chunk_size <- ceiling(n_pars/n_chunks)
  chunk_list <- list()
  for(i in 1:n_chunks){
    if((chunk_size*(i - 1) + 1) <= n_pars){
      chunk_list[[i]] <- draws[,,(chunk_size*(i - 1) + 1):(min(c(chunk_size*i,n_pars)))]
    }
  }
#  summary_list <- parallel::mclapply(chunk_list, posterior::summarise_draws, mc.cores = n_cores)
  summary_list <- lapply(chunk_list, posterior::summarise_draws)
  output <- do.call("rbind", summary_list)
  return(output)
}
```
is radically faster (~ 4x speedup for 1e+5 parameters) than plain old `summarise_draws()`, even though it doesn't implement any parallelization at all.  Note that I haven't taken the time to optimize the default `n_chunks` value (`ceiling(n_pars/500)`) so further speedup is possibly attainable.

### Here's why `summarise_draws()` is slow:

If we take the code for `summarise_draws()` and modify the main for loop as follows:
```
  for (v in variables[i:(i+100)]) {
    draws <- posterior:::drop_dims(x[, , v], dims = 3)
    args <- c(list(draws), .args)
    for (m in names(funs)) {
      out[[v]][[m]] <- do.call(funs[[m]], args)
      if (rlang::is_named(out[[v]][[m]])) {
        # use returned names to label columns
        out[[v]][[m]] <- rbind(out[[v]][[m]])
      }
    }
    out[[v]] <- do.call(cbind, out[[v]])
  }
  ```
  then for a draws object with 2.5e+5 parameters, the above loop executes 10x faster for `i==1` than it does for `i==100000`. So the issue has to do with writing elements deep inside a gargantuan list. 
  
We can squash this difference by preallocating `out` as a matrix and then using the for-loop to fill it, but this breaks for functions like `quantile2` whose outputs are not scalars (I have confirmed that the speed is satisfactory if we remove `quantile2` from the list of functions).  Importantly, we need anticipate any other multi-valued non-default functions that other users might be passing here. One workaround would be to run the summary functions for just the first variable, then parse that output to determine `ncol` of the output matrix as well as wrangle the column indices properly inside the for-loop.

I think I can do that easily enough, but before I do it seems worthwhile to get feedback on the design:
1. Given that `parsummarise_draws` above is simple, reasonably efficient, no work to implement, and lends itself to parallelization, is it worth troubleshooting this at a deeper level inside `summarise_draws` (with our without using `data.table`)?  If not, then I suggest re-naming the current `summarise_draws` function as `summarise_draws_internal` or something, and then writing a new `summarise_draws` which is just a wrapper around the internal part.
2. Whether or not we implement a fix inside the current `summarise_draws`, parallelization would be nice to have.  Doing so requires that we choose a package for implementing the parallelization.  Upthread there's some support for using `future`. 
 Does that sound right to everybody?  @saudiwin (see https://github.com/stan-dev/posterior/issues/124) and I independently used `parallel`, but at least in my case this was without any real thought.
 
 If a fix inside the current `summarise_draws` is desired, then I'll have a few additional questions that I can hold off on for now.

**Comment by paul-buerkner** [comment]
Thank you so much for looking into it! Personally, I am fine with the chunking options for now. And nothing prevents us from implementing an even more efficient version of this later if necessary for some reason. I do think that the "check output size for first variable" and then preparing an output matrix option could also be sensible. Did you run some speed comparisons of these two options as well? I would be hesitant to use `data.table` because this adds yet another big dependency to posterior that I would like to avoid if possible.

With regard to the parallelization option, `future` sounds nice and general to me and I have made good experience with it myself. A simple `parallel` solution is fine for me as well as long as it works on all relevant operating systems (some function in `parallel` don't work on windows).

**Comment by jsocolar** [comment]
Just to mention, I'm testing a (soon-to-be) PR that uses a pre-allocated matrix.  I realized that if we went for the chunking option, setting the n_chunks option would be unnecessarily confusing for users who parallelize.  And if anybody ever tries to massively parallelize something (i.e. with more cores than the default n_chunk) we'd need to write a version that can handle that too.  So I think handling this and #124 separately is the way to go.
